File: Neural Networks/src/Layers/PoolingLayer.java
package Layers;

import Utilities.Matrix;
import Utilities.Pair;

public class PoolingLayer extends Layer {
	private Matrix[] gradients;
	private Matrix[] output;
	
	private int size, stride;
	
	public PoolingLayer(NeuralNetwork network, int layer, int size, int stride) {
		super(network, layer);
		
		this.size = size;
		this.stride = stride;
	}
	
	@Override
	public void initialize() {
		Layer previousLayer = getPreviousLayer();
		
		if(previousLayer == null) {
			inputMatrixCount = network.getInputMatrixCount();
			inputNRows = network.getInputNRows();
			inputNCols = network.getInputNCols();
		} else {
			inputMatrixCount = previousLayer.getOutputMatrixCount();
			inputNRows = previousLayer.getOutputNRows();
			inputNCols = previousLayer.getOutputNCols();
		}
		
		double outputCols = ((inputNCols - size) / ((double) stride)) + 1.0;
		double outputRows = ((inputNRows - size) / ((double) stride)) + 1.0;
		
		if((int) outputCols != outputCols) {
			System.err.println("Invalid Pooling Width!");
			System.exit(-1);
		}
		
		if((int) outputRows != outputRows) {
			System.err.println("Invalid Pooling Height!");
			System.exit(-1);
		}
		
		this.outputNCols = (int) outputCols;
		this.outputNRows = (int) outputRows;
		this.outputMatrixCount = inputMatrixCount;
		
		output = new Matrix[outputMatrixCount];
		for(int i = 0; i < output.length; i++)
			output[i] = new Matrix(outputNRows, outputNCols);
	}
	
	@Override
	public Matrix[] forward(Matrix[] input) {
		for(int i = 0; i < input.length; i++)
			output[i].clear();
		
		if(gradients == null) {
			gradients = new Matrix[input.length];
			for(int i = 0; i < input.length; i++)
				gradients[i] = new Matrix(input[0].getNRows(), input[0].getNCols());
		} else {
			for(int i = 0; i < gradients.length; i++) {
				gradients[i].clear();
			}
		}
		
		for(int i = 0; i < output.length; i++) {
			for(int row = 0; row <= input[0].getNRows() - size; row += stride) {
			for(int col = 0; col <= input[0].getNCols() - size; col += stride) {
				Pair<int[], Double> pair = input[i].mostPositive(row, col, row + size - 1, col + size - 1);
				int[] location = pair.getKey();
				double value = pair.getValue();
				
				gradients[i].set(location[0], location[1], value);
				
				output[i].set(row / stride, col / stride, value);
			}}
		}
		
		return output;
	}
	
	@Override
	protected Matrix[] back(Matrix[] error) {
		return gradients;
	}
}

File: src/LayeredNetwork/Layers/PoolingLayer.cpp
#include "LayeredNetwork/Layers/PoolingLayer.hpp"

PoolingLayer::PoolingLayer(NetworkInformation& networkInformation, int index, int size) : Layer(networkInformation, Pool, index) { 
    this->size = size;
}

void PoolingLayer::initialize() {
    if(inputNRows % size != 0)
        throw std::invalid_argument("Invalid size for pooling layer");

    output = (Matrix*) malloc(sizeof(Matrix) * inputMatrixCount);
    layerGradient = (Matrix*) malloc(sizeof(Matrix) * inputMatrixCount);
    poolIndecies = (Matrix*) malloc(sizeof(Matrix) * inputMatrixCount);

    numPoolIndecies = (inputNRows / size) * (inputNCols / size);

    for(int i = 0; i < inputMatrixCount; i++) {
        new (&output[i]) Matrix(inputNRows / size, inputNCols / size);
        new (&layerGradient[i]) Matrix(inputNRows, inputNCols);
        new (&poolIndecies[i]) Matrix(numPoolIndecies, 2);
    }

    outputMatrixCount = inputMatrixCount;
    outputNRows = inputNRows / size;
    outputNCols = inputNCols / size;
}

void PoolingLayer::toFile(FILE* file) {
    
}

void PoolingLayer::fromFile(FILE* file) {
    
}

Matrix* PoolingLayer::feedForward(Matrix* input) {
    for(int i = 0; i < inputMatrixCount; i++) {

        int index = 0;
        for(int row = 0; row <= inputNRows - size; row += size) {
        for(int col = 0; col <= inputNCols - size; col += size) {
            int positionRow = row;
            int positionCol = col;

            for(int r = row; r < row + size; r++) {
            for(int c = col; c < col + size; c++) {
                if((&input[i])->at(r, c) > (&input[i])->at(positionRow, positionCol)){
                    positionRow = r;
                    positionCol = c;
                }
            }}

            (&output[i])->set(row / size, col / size, (&input[i])->at(positionRow, positionCol));

            (&poolIndecies[i])->set(index, 0, positionRow);
            (&poolIndecies[i])->set(index, 1, positionCol);

            // poolIndecies[i][index][0] = positionRow;
            // poolIndecies[i][index][1] = positionCol;
            index++;
        }}
    }

    return output;
}

Matrix* PoolingLayer::backpropogate(Matrix* error) {
    for(int i = 0; i < outputMatrixCount; i++) {
        (&layerGradient[i])->clear();// could look at the previous positions to set them only
        
        for(int index = 0; index < numPoolIndecies; index++) {
            (&layerGradient[i])->set((&poolIndecies[i])->at(index, 0), (&poolIndecies[i])->at(index, 1), (&error[i])->getData()[index]);
        }
    }

    return layerGradient;
}